WalletWasabi/Blockchain/TransactionBuilding/BnBManager.cs

```diff
@@ -0,0 +1,39 @@
+using System;
```

Comment:
We use `file-scoped` namespaces now, so it should be:

```cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WalletWasabi.Blockchain.TransactionOutputs;

namespace WalletWasabi.Blockchain.TransactionBuilding;

public static class BnBManager
{
	public static bool CanCreateChangelessTransaction(List<SmartCoin> availableCoins, long target, [NotNullWhen(true)] out List<SmartCoin>? selectedCoins)
	{
		selectedCoins = null;
		var longs = availableCoins.Select(x => x.Amount.Satoshi).ToList();
		BranchAndBound branchAndBound = new(longs);

		if (branchAndBound.TryGetExactMatch(target, out List<long>? bnbResult))
		{
			selectedCoins = new();
			foreach (var coin in availableCoins)
			{
				foreach (var item in bnbResult)
				{
					if (coin.Amount == item)
					{
						selectedCoins.Add(coin);
						bnbResult.Remove(item);
						break;
					}
				}
			}
			return true;
		}
		return false;
	}
}
```

Reply:
Anyway, the code in `BnBManager` is not correct because you cannot spend `SmartCoin.Amount` really. You can spend only an effective value of a coin (i.e. coin value - fees given by fee rate). We have `NBitcoinExtensions.EffectiveValue(...)` functions to get you these effective values. Based on this, I would rewrite the code (NOT TESTED!) like this:

```cs
using NBitcoin;
using System.Collections.Generic;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using WalletWasabi.Blockchain.TransactionOutputs;

namespace WalletWasabi.Blockchain.TransactionBuilding;

public static class ChangelessTransactionCoinSelector
{
	public static bool TryGetCoins(List<SmartCoin> availableCoins, FeeRate feeRate, long target, [NotNullWhen(true)] out List<SmartCoin>? selectedCoins)
	{
		selectedCoins = null;

		// Keys are effective values of smart coins in satoshis.
		SortedDictionary<long, SmartCoin> inputs = new(availableCoins.ToDictionary(x => x.EffectiveValue(feeRate).Satoshi, x => x));

		// Pass smart coins' effective values in ascending order.
		BranchAndBound branchAndBound = new(inputs.Keys.ToList());

		if (branchAndBound.TryGetExactMatch(target, out List<long>? solution))
		{
			// TODO: We should probably sort in `TryGetExactMatch` right away.
			solution.Sort();

			selectedCoins = new();
			int i = 0;

			foreach ((long effectiveSatoshis, SmartCoin smartCoin) in inputs)
			{
				if (effectiveSatoshis == solution[i])
				{
					i++;
					selectedCoins.Add(smartCoin);
				}
			}

			Debug.Assert(i == solution.Count - 1);

			return true;
		}

		return false;
	}
}
```

And this PR should contain tests for this `ChangelessTransactionCoinSelector` class so that we are confident it works OK. 

Also one has to make sure that fee for the single transaction input is paid (I think) and I'm not sure whether `long target` is with or without fee. So what I'm saying is that the code above should be taken with a grain of salt but it's still better than the original code.

Reply:
@lontivero Could you please have a look at https://github.com/zkSNACKs/WalletWasabi/pull/6999#discussion_r780980972 whether it makes sense to you?

Reply:
```
SortedDictionary<long, SmartCoin> inputs = new(availableCoins.ToDictionary(x => x.EffectiveValue(feeRate).Satoshi, x => x));
```
 This line thrown an exception, because I have more than one coin with unique value:
```
System.ArgumentException: 'An item with the same key has already been added. Key: 65467'
```
Tried to swtich up `long` and `SmartCoin`, but SortedDictionary can't sort with `SmartCoin`, so it threw this exception:
```
System.ArgumentException: 'At least one object must implement IComparable.'
```
Any ideas?

Reply:
We are working on a fix, maybe sorting availableCoins before making it into a dictionary.

Reply:
I would try to switch from `SortedDictionary<long, SmartCoin>` to `List<KeyValuePair<long, SmartCoin>>()` and sort the list by the longs.